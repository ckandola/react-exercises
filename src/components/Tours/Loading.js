import React from 'react';

const Loading = () => {
    return (
        <div className="tour-loading">
            <h1>loading...</h1>
        </div>
    );
};

export default Loading;
