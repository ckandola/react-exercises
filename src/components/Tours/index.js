// from https://www.youtube.com/watch?v=a_7Z7C_JCyo

import React, {useState, useEffect} from 'react';
import Loading from './Loading';
import Tour from './Tour';
import './index.css';

const url = 'https://course-api.com/react-tours-project';

const Tours = () => {
    const [loading, setLoading] = useState(true);
    const [tours, setTours] = useState([]);

    const fetchTours = async () => {
        setLoading(true);
        try {
            const response = await fetch(url);
            const tours = await response.json();
            setLoading(false);
            setTours(tours);
        } catch (error) {
            setLoading(false);
            console.log(error);
        }
    };

    const deleteTour = id => {
        const newTours = tours.filter(tour => {
            return tour.id !== id;
        })
        setTours(newTours);
    }

    useEffect(() => {
        fetchTours();
    }, []);

    if (loading) {
        return (
            <main className="tours-main"><Loading /></main>
        );
    }

    return (
        <section>
            <div className="title">
                {tours.length !== 0 ? 
                    <div>
                        <h2>Our Tours</h2>
                        <div className="underline"></div>
                        {tours.map(tour => {
                            return <Tour key={tour.id} {...tour} deleteTour={deleteTour}></Tour>
                        })}
                    </div> :
                    <div>
                        <h2>No Tours Left</h2>
                        <button className="btn" onClick={fetchTours}>Refresh</button>
                    </div>
                }
            </div>
        </section>
    );
};

export default Tours;
