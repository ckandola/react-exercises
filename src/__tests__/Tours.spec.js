import React from 'react';
import {render, waitFor, act} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Tours from '../components/Tours';

describe('Tours tests', () => {
    it('matches the snapshot', () => {
        const tours = renderer.create(<Tours />);
        const tree = tours.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should display all tour info after loading', async () => {
        const view = render(<Tours />);
        await act(async () => {
            await waitFor(() => expect(view.container.getElementsByClassName('single-tour').length).toBe(5));
        });
    });

    it('removes a Tour from the list when "Not Interested" is clicked', async () => {
        const view = render(<Tours />);
        await act(async () => {
            await waitFor(() => expect(view.container.getElementsByClassName('tour-delete-btn').length).toBe(5));
            userEvent.click(view.container.getElementsByClassName('tour-delete-btn')[0]);
            expect(view.container.getElementsByClassName('tour-delete-btn').length).toBe(4);
        });

    });
});

