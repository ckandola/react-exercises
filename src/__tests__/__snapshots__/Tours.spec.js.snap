// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tours tests matches the snapshot 1`] = `
<main
  className="tours-main"
>
  <div
    className="tour-loading"
  >
    <h1>
      loading...
    </h1>
  </div>
</main>
`;
